add single execution option for ConversationalRetrievalChain

jpzhangvincent opened the pull request:
# add single execution option for ConversationalRetrievalChain

<!--
Thank you for contributing to LangChain! Your PR will appear in our next release under the title you set. Please make sure it highlights your valuable contribution.

Replace this with a description of the change, the issue it fixes (if applicable), and relevant context. List any dependencies required for this change.

After you're done, someone will review your PR. They may suggest improvements. If no one reviews your PR within a few days, feel free to @-mention the same people again, as notifications can get lost.
-->

- The current default workflow for ConversationalRetrievalChain is to rephrase the question based on the chat history and then use the rephrased question to perform retrieval and question answering. 
- This current two-step process has pros and cons. It helps address the context size limit but it could increase unnecessary latency and incur potential information loss. 
- We can make the `question_generator` (chain) argument optional to allow it to bypass the rephrasing step and use a custom prompt to combine  the `chat_history`, `context` and `question` together for single execution. This could give users more flexibility based on their use cases and requirements.

<!-- Remove if not applicable -->

Fixes #5123 

## Before submitting

<!-- If you're adding a new integration, include an integration test and an example notebook showing its use! -->
I updated the notebook to include the `## ConversationalRetrievalChain with only custom combine_docs_chain` section

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:
@hwchase17 @dev2049

<!-- For a quicker response, figure out the right person to tag with @

        @hwchase17 - project lead

        Tracing / Callbacks
        - @agola11

        Async
        - @agola11

        DataLoaders
        - @eyurtsev

        Models
        - @hwchase17
        - @agola11

        Agents / Tools / Toolkits
        - @vowelparrot
        
        VectorStores / Retrievers / Memory
        - @dev2049
        
 -->

jpzhangvincent committed:
fix linting and update notebook

jpzhangvincent committed:
fix with ruff

jpzhangvincent commented:
I believe this feature provides practical values. Can I get a review on this one? @hwchase17 @dev2049

franco-roura commented:
Hello guys, we're having a similar issue with the `ConversationalRetrievalQAChain` class. 
When adding a Pinecone vector store for memory, it can even reach 45 seconds of total response time, which is far from ideal. This PR does seem to have the potential to mitigate the issue, is it being actively looked at?

Thanks in advance.

dev2049 commented on langchain/chains/conversational_retrieval/base.py:
```diff
@@ -107,6 +110,12 @@ def _call(
+            and "context"
+            not in self.combine_docs_chain.llm_chain.prompt.input_variables
+        ):
+            self.combine_docs_chain.llm_chain.prompt = CHAT_RETRIEVAL_QA_PROMPT
```
why do this at runtime, why not set the prompt of combine_docs_chain when instantiating the ConversationalRetrievalChain?

jpzhangvincent replied:
@dev2049 can you review again?

dev2049 reviewed (commented):
making question_generator optional makes sense to me! but don't think we should be messing with prompts at runtime, you should be able to set those when instantiating the chain in first place

jpzhangvincent committed:
move the default combined prompt logic to the initialization

jpzhangvincent committed:
fix typing

jpzhangvincent committed:
fix typing

jpzhangvincent committed:
fix linting length

hwchase17 commented:
i think we need just better documentation around conversational retrieval chains in general

jpzhangvincent commented:
> i think we need just better documentation around conversational retrieval chains in general

I have updated the notebook to show how to set `question_generator` as None and use a custom prompt. But feel free to let me know anything else needed to get this PR merged. Thanks!

jpzhangvincent commented:
Hoping this can get merged soon since I need this patch for a work project! @hwchase17 @dev2049 any feedback?

italojs commented:
👀👀👀

jpzhangvincent commented:
@dev2049 @hwchase17 can you take a look at this PR? Sorry for chasing..

anthonycoded commented:
How can I implement this solution in my project?

jpzhangvincent committed:
fix typos

jpzhangvincent commented on langchain/chains/combine_documents/base.py:
```diff
@@ -36,6 +37,7 @@ def format_document(doc: Document, prompt: BasePromptTemplate) -> str:
 class BaseCombineDocumentsChain(Chain, ABC):
     """Base interface for chains combining documents."""
 
+    llm_chain: LLMChain
```
Yes because we set the `prompt` variable like
`combine_docs_chain.llm_chain.prompt = CHAT_RETRIEVAL_QA_PROMPT` in the `validate_combine_docs_chain` function for the error handling and I found I have to put the type there to pass the linting check.

jpzhangvincent replied:
@leferradkw let me know if any other questions! Hope to get the PR merged soon! :)

kowsikgelli commented:
A must needed feature. I am looking for this.

leo-gan commented:
@jpzhangvincent Hi , could you, please, resolve the merging issues and address the last comments (if needed)? After that, ping me and I push this PR for the review. Thanks!

leo-gan commented:
@jpzhangvincent It is again :( Could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!

jpzhangvincent commented:
> @jpzhangvincent It is again :( Could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!

@leo-gan I believe I have fixed the merge conflicts. I tried to add a test or an example in the notebook(if needed) but not sure where/which file to add specifically. Open to suggestion!

leo-gan commented:
@jpzhangvincent There are failed unit tests and pydantic compatibility tests. When you fix them we will be ready.

jpzhangvincent committed:
fix unit test

jpzhangvincent committed:
fixes and update doc

jpzhangvincent committed:
fix linting

jpzhangvincent commented:
@leo-gan I believe I had resolved the test issues.. please review again and feel free to make code changes directly!

techcentaur commented:
@dev2049 @baskaryan can we get a merge for this? sorry for jumping into a conversation. But waiting to use this feature asap in Prod code.

leferradkw reviewed (approved):
LGTM!

leo-gan commented:
@baskaryan Could you, please, review it? Looks like it urgent for users.

jpzhangvincent commented:
So any update on this PR? Thanks. - @leo-gan @baskaryan

pixeldu commented:
slow for me as well. 30 seconds

baskaryan commented on libs/langchain/langchain/chains/conversational_retrieval/base.py:
```diff
@@ -112,6 +115,27 @@ def output_keys(self) -> List[str]:
         return _output_keys
 
+    @root_validator
+    def validate_combine_docs_chain(cls, values: Dict) -> Dict:
```
chat_history is passed into the combine_docs_chain whether or not there's a question_generator. think it's confusing if we change prompts for users, think it's clearer if things error when prompt doesn't support chat_history var

yuanpusheng commented:
When will it be fixed？This problem is very urgent。

DevOtts commented:
just adding myself to the chorus of those begging for this PR to be merged @leo-gan @baskaryan

jpzhangvincent committed:
fix conflict

jpzhangvincent committed:
fix conflict

jpzhangvincent committed:
fix conflict

jpzhangvincent committed:
fix conflict

jpzhangvincent committed:
revert notebook changes

jpzhangvincent committed:
revert notebook to be same with master

jpzhangvincent committed:
fix ruff errors

jpzhangvincent committed:
fix previous changes

jpzhangvincent committed:
add unit test

jpzhangvincent commented:
@baskaryan @leo-gan Okay I have addressed the feedback and added a unit test. Hopeful we can get this merged soon since it could be low-hanging fruit but useful feature.

jpzhangvincent committed:
fix formatting

jpzhangvincent committed:
fix test file  formatting

jpzhangvincent committed:
Merge branch 'master' into add-convQARetrchain-single-run

jpzhangvincent committed:
fix pytest error

jpzhangvincent committed:
Merge branch 'master' into add-convQARetrchain-single-run

jpzhangvincent commented:
> @baskaryan @leo-gan Okay I have addressed the feedback and added a unit test. Hopeful we can get this merged soon since it could be low-hanging fruit but useful feature.

@baskaryan @leo-gan @hwchase17 thoughts on merging this PR?

jpzhangvincent committed:
Merge branch 'master' into add-convQARetrchain-single-run

jpzhangvincent commented:
Can you please review again? Thanks! @baskaryan @leo-gan

ccurme commented:
Thank you for this. Closing because this is readily achievable with [create_retrieval_chain](https://api.python.langchain.com/en/latest/chains/langchain.chains.retrieval.create_retrieval_chain.html). See the example snippet in the API reference. You can also reference [this guide](https://python.langchain.com/v0.2/docs/how_to/migrate_chains/#conversationalretrievalchain) on migrating from ConversationalRetrievalChain to LCEL implementations. Please let me know if you have any other concerns.
